packages/grafana-data/src/field/fieldOverrides.ts

```diff
@@ -290,10 +290,42 @@ export function setDynamicConfigValue(config: FieldConfig, value: DynamicConfigV
   }
 }
 
+/**
+ * Merge the links from the ds config and the panel config
+ * @param configLinks coming from the datasource
+ * @param defaultLinks coming from the panel config
+ * @returns unique merged links
+ */
+
+function mergeLinks(configLinks: DataLink[], defaultLinks: DataLink[]) {
+  // Combine the config links and the default links
+  const combinedLinks = [...configLinks, ...defaultLinks];
+
+  // Deduplicate the combined array of links
+  // note: this is necessary to prevent default links from being displayed multiple times
+  // as setFieldConfigDefaults is called multiple times for each data frame
+
+  const uniqueLinks = combinedLinks.reduce((accumulator: DataLink[], currentLink) => {
+    const existingLink = accumulator.find((link) => JSON.stringify(link) === JSON.stringify(currentLink));
+    if (!existingLink) {
+      accumulator.push(currentLink);
+    }
+
+    return accumulator;
+  }, []);
+
+  return uniqueLinks;
+}
+
 // config -> from DS
 // defaults -> from Panel config
 export function setFieldConfigDefaults(config: FieldConfig, defaults: FieldConfig, context: FieldOverrideEnv) {
   for (const fieldConfigProperty of context.fieldConfigRegistry.list()) {
+    // For cases where we have links on the datasource config and the panel config, we need to merge them
+    if (config.links && defaults.links) {
+      config.links = mergeLinks(config.links, defaults.links);
+    }
```

Comment:
Why is this done within the loop? It's something that doesn't seem like fieldConfigProperty dependant.

Reply:
It looks like this might be the reason you need to deduplicate the links in merge. `setFieldConfigDefaults` is called once per each field, so I don't see any other reason why links would be duplicated there.

Reply:
I was checking this also. I see also the function `applyFieldConfigDefaults`, which I think is meant to do the merging between the passed config and the default one (correct me if I'm wrong).

So, maybe that's the place to do this operation. WDYT?

Reply:
No, I don't think so. That function is building the field configuration _of the panel_. What we are looking for here is the _actual application of that field config on Fields_. Panel's field config is considered a source that is later on applied to particular fields. So there are two sources of the field config - it's Panel and the raw data directly returned from the data source. And these are not known to each other until `PQR.getData`.

Reply:
@dprokop good catch! I missed completely the loop 🙈